metrics/ci_integrations/lib/source/github_actions/adapter/github_actions_source_client_adapter.dart

```diff
@@ -0,0 +1,326 @@
+      status: RunStatus.completed,
+      perPage: perPageLimit,
+      page: 1,
+    );
+
+    final List<WorkflowRun> result = [];
+
+    while (page.hasNextPage) {
+      final runs = page.values;
+
+      for (final run in runs) {
+        if (run.number > lastBuildNumber) {
+          result.add(run);
+        }
+        if (run.number == lastBuildNumber) return result;
+      }
+
+      page = await _fetchNextRunsPage(page);
+    }
+
+    return result;
+  }
+
+  /// Fetches the [WorkflowRunDuration] of the given [run].
+  Future<WorkflowRunDuration> _fetchRunDuration(WorkflowRun run) async {
+    final durationFetchResult =
+        await githubActionsClient.fetchRunDuration(run.id);
+
+    _throwIfInteractionUnsuccessful(durationFetchResult);
+
+    return durationFetchResult.result;
+  }
+
+  /// Fetches the coverage after the given [run].
+  /// If the coverage file was not found, returns 'null'.
+  Future<Percent> _fetchCoverage(WorkflowRun run) async {
+    WorkflowRunArtifactsPage page = await _fetchArtifactsPage(
+      runId: run.id,
+      perPage: perPageLimit,
+      page: 1,
+    );
+
+    while (true) {
+      final artifacts = page.values;
+
+      final coverageArtifact = artifacts.firstWhere(
+        (artifact) => artifact.name == 'coverage-summary.json',
+        orElse: () => null,
+      );
+
+      if (coverageArtifact != null) {
+        return _mapArtifactToCoverage(coverageArtifact);
+      }
+
+      if (!page.hasNextPage) return null;
+
+      page = await _fetchNextArtifactsPage(page);
+    }
+  }
+
+  /// Fetches a [WorkflowRunsPage] with the given [workflowIdentifier].
+  ///
+  /// A [workflowIdentifier] is either a workflow id or a name of the file
+  /// that defines the workflow.
+  ///
+  /// A [status] is used as a filter query parameter to define the
+  /// [WorkflowRun.status] of runs to fetch.
+  ///
+  /// A [perPage] is used for limiting the number of runs and pagination in pair
+  /// with the [page] parameter.
+  ///
+  /// A [page] is used for pagination and defines a page of runs to fetch.
+  /// If the [page] is `null` or omitted, the first page is fetched.
+  Future<WorkflowRunsPage> _fetchRunsPage({
+    String workflowIdentifier,
+    RunStatus status,
+    int perPage,
+    int page,
+  }) async {
+    final runsInteraction = await githubActionsClient.fetchWorkflowRuns(
+      workflowIdentifier,
+      status: status,
+      perPage: perPage,
+      page: page,
+    );
+
+    _throwIfInteractionUnsuccessful(runsInteraction);
+
+    return runsInteraction.result;
+  }
+
+  /// Fetches a [WorkflowRunArtifactsPage] of a run with the given [runId].
+  ///
+  /// A [perPage] is used for limiting the number of runs and pagination in pair
+  /// with the [page] parameter.
+  ///
+  /// A [page] is used for pagination and defines a page of runs to fetch.
+  /// If the [page] is `null` or omitted, the first page is fetched.
+  Future<WorkflowRunArtifactsPage> _fetchArtifactsPage({
```

Comment:
Should we inline this method? It seems like it only used once and only delegates call.